File: sdks/client/go/lenny/types_test.go
// SPDX-License-Identifier: MIT

package lenny

import (
	"encoding/json"
	"testing"
)

// spec: §7.1 sessionIsolationLevel — the create response carries a
// conversationContinuity field whose value-set is platform|none keyed
// off the executionMode session|service value-set. The SDK decodes both
// fields verbatim from the wire envelope.
func TestIsolationLevelDecodesConversationContinuity(t *testing.T) {
	cases := []struct {
		name         string
		wire         string
		wantMode     string
		wantContinue string
	}{
		{
			name: "session mode preserves platform continuity",
			wire: `{
				"id": "sess_1",
				"sessionIsolationLevel": {
					"executionMode": "session",
					"isolationProfile": "gvisor",
					"podReuse": false,
					"residualStateWarning": false,
					"conversationContinuity": "platform"
				}
			}`,
			wantMode:     "session",
			wantContinue: "platform",
		},
		{
			name: "service mode declares no continuity",
			wire: `{
				"id": "sess_2",
				"sessionIsolationLevel": {
					"executionMode": "service",
					"isolationProfile": "runc",
					"podReuse": true,
					"scrubPolicy": "none",
					"residualStateWarning": true,
					"conversationContinuity": "none"
				}
			}`,
			wantMode:     "service",
			wantContinue: "none",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			var got CreateSessionResult
			if err := json.Unmarshal([]byte(tc.wire), &got); err != nil {
				t.Fatalf("Unmarshal: %v", err)
			}
			if got.IsolationLevel.ExecutionMode != tc.wantMode {
				t.Errorf("ExecutionMode = %q, want %q", got.IsolationLevel.ExecutionMode, tc.wantMode)
			}
			if got.IsolationLevel.ConversationContinuity != tc.wantContinue {
				t.Errorf("ConversationContinuity = %q, want %q", got.IsolationLevel.ConversationContinuity, tc.wantContinue)
			}
		})
	}
}

// spec: §7.1 sessionIsolationLevel — a wire envelope that omits
// conversationContinuity decodes to the empty string rather than
// failing, matching the optional decode of every other isolation field.
func TestIsolationLevelOmittedConversationContinuity(t *testing.T) {
	var got CreateSessionResult
	if err := json.Unmarshal([]byte(`{"sessionIsolationLevel":{"executionMode":"session"}}`), &got); err != nil {
		t.Fatalf("Unmarshal: %v", err)
	}
	if got.IsolationLevel.ConversationContinuity != "" {
		t.Errorf("ConversationContinuity = %q, want empty", got.IsolationLevel.ConversationContinuity)
	}
}

File: sdks/client/typescript/test/types.test.mjs
// SPDX-License-Identifier: MIT

// Unit tests for the §7.1 sessionIsolationLevel decode of the
// TypeScript client SDK. The tests import the built ESM bundle (the
// dual-build output the package publishes) and stand up a local
// node:http server that returns a POST /v1/sessions response carrying
// the sessionIsolationLevel object.
//
// The tests cover the conversationContinuity field across the session
// and service execution modes: createSession surfaces it verbatim so a
// consumer can branch on platform vs none continuity.

import assert from 'node:assert/strict';
import { createServer } from 'node:http';
import test from 'node:test';

import { newClient } from '../dist/esm/index.js';

// startServer starts an HTTP server bound to a loopback port and
// returns its base URL plus a close function.
function startServer(handler) {
  return new Promise((resolve) => {
    const server = createServer(handler);
    server.listen(0, '127.0.0.1', () => {
      const { port } = server.address();
      resolve({
        url: `http://127.0.0.1:${port}`,
        close: () => new Promise((done) => server.close(done)),
      });
    });
  });
}

// createSessionReturning answers POST /v1/sessions with the supplied
// sessionIsolationLevel envelope and returns the decoded result.
async function createSessionReturning(isolationLevel) {
  const srv = await startServer((req, res) => {
    res.setHeader('Content-Type', 'application/json');
    res.end(
      JSON.stringify({
        id: 'sess_1',
        state: 'created',
        uploadToken: 'tok',
        sessionIsolationLevel: isolationLevel,
      }),
    );
  });
  try {
    const client = newClient(srv.url);
    return await client.createSession({ runtimeRef: 'rt' });
  } finally {
    await srv.close();
  }
}

// spec: §7.1 sessionIsolationLevel — session mode reports platform
// conversation continuity alongside the session executionMode value.
test('createSession decodes platform continuity for session mode', async () => {
  const result = await createSessionReturning({
    executionMode: 'session',
    isolationProfile: 'gvisor',
    podReuse: false,
    residualStateWarning: false,
    conversationContinuity: 'platform',
  });
  assert.equal(result.sessionIsolationLevel.executionMode, 'session');
  assert.equal(result.sessionIsolationLevel.conversationContinuity, 'platform');
});

// spec: §7.1 sessionIsolationLevel — service mode reports no
// conversation continuity alongside the service executionMode value.
test('createSession decodes none continuity for service mode', async () => {
  const result = await createSessionReturning({
    executionMode: 'service',
    isolationProfile: 'runc',
    podReuse: true,
    scrubPolicy: 'none',
    residualStateWarning: true,
    conversationContinuity: 'none',
  });
  assert.equal(result.sessionIsolationLevel.executionMode, 'service');
  assert.equal(result.sessionIsolationLevel.conversationContinuity, 'none');
});
